geps/gep-1713/index.md

```diff
@@ -0,0 +1,377 @@
+
+(See status definitions [here](overview.md#status).)
+
+## tl;dr
+
+The Gateway Resource is a contention point since it is the only place to attach listeners with certificates. We propose a mechanism to allow distinct Gateway resources to be logically merged.
+
+## Goals
+
+- Define a mechanic to merge multiple Gateways (logically)
+- Define a set of acceptable properties that can be merged and their semantics
+
+## Non-Goals
+
+- Apply a Gateway resource onto N distinct gateways (one to many)
+
+## Introduction
+
+Knative generates on demand per-service certificates using HTTP-01 challenges. 
+There can be O(1000) Knative Services in the cluster which means we have O(1000) distinct certificates. 
+Thus updating a single Gateway resource with this many certificates is a contention point and inhibits horizontal scaling of our controllers.
+[Istio Ambient](https://istio.io/v1.15/blog/2022/introducing-ambient-mesh/), similarly, creates a listener per Kubernetes service.
+
+More broadly, large scale gateway users often expose O(1000) domains, but are currently limited by the maximum of 16 `listeners`.
+
+The spec currently has language to indicate implementations `MAY` merge Gateways resources but the mechanic isn't defined.
+https://github.com/kubernetes-sigs/gateway-api/blob/541e9fc2b3c2f62915cb58dc0ee5e43e4096b3e2/apis/v1beta1/gateway_types.go#L76-L78
+
+## API
+
+This proposal has two aspects: configuration for the parent Gateway, and configuration for the child Gateway.
+
+A "parent" Gateway _does not_ reference another Gateway. A parent Gateway MUST explicitly opt into merging.
+This is done with a new field `allowedChildren` in the `spec.infrastructure` of a Gateway.
+If, and only if, they have done so, the same Gateway is also permitted to leave `listeners` empty (currently, there is a `MinItems=1` restriction).
+
+A "child" Gateway references a parent Gateway. 
+This is done with a new field `attachTo` in the `spec.infrastructure` stanza of a Gateway.
+The `attachTo` field is a new type `GatewayObjectReference`.
+Although the use of `GatewayObjectReference` allows users to attach to any `kind`, this GEP only defines the behavior of attaching a Gateway to another Gateway.
+
+A "sibling" is a Gateway that shares a parent with another child Gateway.
+
+Status requirements are specified [below](#status-fields).
+
+See [GEP-1867](https://github.com/kubernetes-sigs/gateway-api/pull/1868) for more use cases of `infrastructure`.
+
+
+#### Go
+
+```go
+type GatewayInfrastructure struct {
+	// ...
+
+	// AllowedChildren allows child objects to attach to this Gateway.
+	// A common scenario is to allow other objects to add listeners to this Gateway.
+	AllowedChildren *AllowedChildren `json:"allowedChildren,omitempty"`
+
+	// AttachTo allows the Gateway to associate itself with another resource.
+	// A common scenario is to reference another Gateway which marks
+	// this Gateway a child of another.
+	AttachTo GatewayObjectReference `json:"attachTo"`
```

Comment:
Not sure if this was discussed before, but would there be any appetite for this being a list so a single Gateway could attach to multiple parents? One way this could be useful is if there were separate parent/root Gateways for say prod/staging and any delegated child Gateways could attach to both to avoid duplicating config.

I personally don't have a particular affinity for this, just an idea in case it is useful/something to consider, I don't want to make this resource relationship too complicated if it is not useful

Has implications for the gatewayclass name discussion below as well

Reply:
It would make status fairly challenging. I think this can be solved at higher levels as well (tooling like helm, etc) so I am not sure its worth it

Reply:
We considered something similar during the route delegation exploration in https://github.com/kubernetes-sigs/gateway-api/issues/1058, but I think trying to add this here could significantly scope creep this GEP.

Reply:
Going to resolve this - as this idea introduces _more_ complexity for a GEP which folks thing is already complex ;)